cql3/multi_column_relation.hh

```diff
@@ -145,7 +145,10 @@ protected:
             return cs->column_specification;
         });
         auto t = to_term(col_specs, *get_value(), db, schema->ks_name(), bound_names);
-        return ::make_shared<restrictions::multi_column_restriction::EQ>(schema, rs, t);
+        auto restr = ::make_shared<restrictions::multi_column_restriction::EQ>(schema, rs, t);
+        using namespace restrictions::wip;
+        restr->wip_equivalent = ::make_shared<expression>(binary_operator{rs, _relation_type, t});
```

Comment:
Is unique_ptr viable here? I haven't verified but I'm guessing that expressions from a tree, not a DAG, so shared pointers are not needed.

Reply:
I think it's better to use the literal here instead of _relation_type. A decision was already made to use equality by the caller, we shouldn't assume that decision was based on _relation_type again (although in practice they will always match).

Reply:
Another question is whether we should use separate types for individual binary operators. Equality, inequality, and ordering relations are treated very differently (equality can translate to an index or primary key read, inequality is always filtering, and ordering can translate to filtering or {primary key, index} scan. OTOH the visitor can create perform a switch on the relation type.

Reply:
> Is unique_ptr viable here?

I argued before that it's too early to know.  I'm afraid of a scenario where two `restriction` objects continue to exist independently after one has been merged to the other (which the interface allows); with unique_ptr, one of those will have an invalid wip_equivalent.  The restriction-creation code has a couple more future changes planned, so I'd like to defer the ownership decision until it settles.

> Another question is whether we should use separate types for individual binary operators.

I don't think that would be a big win compared to visitors `switch`ing on the operator_type, unless we grow a deeper hierarchy of operators.

Reply:
> > Is unique_ptr viable here?
> 
> I argued before that it's too early to know. I'm afraid of a scenario where two `restriction` objects continue to exist independently after one has been merged to the other (which the interface allows); with unique_ptr, one of those will have an invalid wip_equivalent. The restriction-creation code has a couple more future changes planned, so I'd like to defer the ownership decision until it settles.

Logically if we merge a condition into another condition we don't need to keep the original.

I can see a situation where we only merge partially, but still the part that gets merged is moved from the source.

And, if it happens that an expression does get used in two places, then we can choose to clone it instead of using a shared_ptr, if we decide it's worth it.

> 
> > Another question is whether we should use separate types for individual binary operators.
> 
> I don't think that would be a big win compared to visitors `switch`ing on the operator_type, unless we grow a deeper hierarchy of operators.

Reply:
The shared_ptr thing is minor. It simplifies the code to have unique_ptr instead, but in the end this isn't make-or-break.